File: javascript/src/formatRupiah.ts
import { SATUAN_LONG } from './constants';

/**
 * Formats a given number to a Rupiah formatted string.
 * @param {number} n
 * @param {number} level
 * @returns {string}
 * @example
 * ```js
 * formatRupiah(-899_999.976); // "(Rp 899.999,976)"
 * formatRupiah(499_500.499);  // "Rp 499.500,499"
 * ```
 */
export function formatRupiah(n: number, level = -1): string {
  if (typeof n !== 'number' && !Number.isNaN(n)) {
    throw new TypeError('isi dari argumen pertama harus berupa angka, bukan string atau lainnya');
  }

  const t = n.toLocaleString('id-ID', { signDisplay: 'never', style: 'decimal' });

  if (level < 0) {
    return `${n < 0 ? '(' : ''}Rp ${t}${n < 0 ? ')' : ''}`;
  }

  const s = t.split('.');
  const v = s.slice(0, s.length - level);
  return `${n < 0 ? '(' : ''}Rp ${v.join('.')} ${SATUAN_LONG[level + 1]}${n < 0 ? ')' : ''}`;
}

File: ruby/lib/manusier/format_rupiah.rb
# frozen_string_literal: true

module Manusier
  def self.format_rupiah(n, level = 0)
    raise(StandardError, "isi dari argumen pertama harus berupa angka, bukan string atau lainnya") unless n.is_number?

    t = private_self_number_to_dotted_string(n)

    if level.zero?
      return "#{n.negative? ? "(" : ""}Rp #{t}#{n.negative? ? ")" : ""}"
    end

    s = t.split(".")
    v = s.slice(0, s.length - level)
    "#{n.negative? ? "(" : ""}Rp #{v.join(".")} #{$SATUAN_LONG[level + 1]}#{n.negative? ? ")" : ""}"
  end

  def self.private_self_split_per_three(str)
    temp = []

    m = str.length % 3
    case m
    when 2
      temp << str.slice!(0, 2)
    when 1
      temp << str.slice!(0, 1)
    end

    i = 0
    while i < str.length
      temp << str.slice(i, i + 3)
      i += 3
    end

    temp
  end

  def self.private_self_number_to_dotted_string(n)
    s = private_self_split_per_three(n.abs.to_s)
    s.join(".")
  end
end
